From 2520149b963f05e22290eafebc69c87e12c481d2 Mon Sep 17 00:00:00 2001
From: Diya Bera <diyab@google.com>
Date: Fri, 3 Apr 2026 19:51:55 +0000
Subject: [PATCH 3/8] Remove EXTRA_DATA from ConfirmDeviceCredentialActivity

Bug: 494585668
Test: N/A
Flag: EXEMPT CVE_FIX
Change-Id: I29eb7981d202c878c51b9e68ab1564d662ec0fc1
(cherry picked from commit e99f4f51fe5823cd990f93984517ee2875809c7f)
DISABLE_TOPIC_PROTECTOR: omitting STS test
---
 src/com/android/settings/Utils.java           | 28 ++-----------------
 .../fingerprint/FingerprintSettings.java      | 25 +++++++++++------
 .../ConfirmDeviceCredentialActivity.java      |  5 +---
 3 files changed, 21 insertions(+), 37 deletions(-)

diff --git a/src/com/android/settings/Utils.java b/src/com/android/settings/Utils.java
index 295608e677d..96fb5934d18 100644
--- a/src/com/android/settings/Utils.java
+++ b/src/com/android/settings/Utils.java
@@ -29,7 +29,6 @@ import static android.text.format.DateUtils.FORMAT_SHOW_DATE;
 import static com.android.settings.password.ConfirmDeviceCredentialActivity.BIOMETRIC_PROMPT_AUTHENTICATORS;
 import static com.android.settings.password.ConfirmDeviceCredentialActivity.BIOMETRIC_PROMPT_HIDE_BACKGROUND;
 import static com.android.settings.password.ConfirmDeviceCredentialActivity.BIOMETRIC_PROMPT_NEGATIVE_BUTTON_TEXT;
-import static com.android.settings.password.ConfirmDeviceCredentialActivity.EXTRA_DATA;
 
 import android.app.ActionBar;
 import android.app.Activity;
@@ -1753,30 +1752,10 @@ public final class Utils extends com.android.settingslib.Utils {
         final UserManager userManager = (UserManager) fragment.getContext().getSystemService(
                 UserManager.class);
         fragment.startActivityForResult(getIntentForBiometricAuthentication(fragment.getResources(),
-                getEffectiveUserId(userManager, userId), hideBackground, null /* data */),
+                getEffectiveUserId(userManager, userId), hideBackground),
                 requestCode);
     }
 
-    /**
-     * Launch biometric prompt for mandatory biometrics. Call
-     * {@link #requestBiometricAuthenticationForMandatoryBiometrics(Context, boolean, int)}
-     * to check if all requirements for mandatory biometrics is satisfied
-     * before launching biometric prompt.
-     *
-     * @param fragment       corresponding fragment of the surface
-     * @param requestCode    for starting the new activity
-     * @param userId         user id for the authentication request
-     * @param hideBackground if the background activity screen needs to be hidden
-     * @param data           additional info to returned to the activity after authentication
-     *                       has ended
-     */
-    public static void launchBiometricPromptForMandatoryBiometrics(@NonNull Fragment fragment,
-            int requestCode, int userId, boolean hideBackground, @Nullable Intent data) {
-        final UserManager userManager = (UserManager) fragment.getContext().getSystemService(
-                UserManager.class);
-        fragment.startActivityForResult(getIntentForBiometricAuthentication(fragment.getResources(),
-                getEffectiveUserId(userManager, userId), hideBackground, data), requestCode);
-    }
 
     /**
      * Launch biometric prompt for mandatory biometrics. Call
@@ -1794,7 +1773,7 @@ public final class Utils extends com.android.settingslib.Utils {
         final UserManager userManager = activity.getSystemService(UserManager.class);
         activity.startActivityForResult(getIntentForBiometricAuthentication(
                 activity.getResources(), getEffectiveUserId(userManager, userId),
-                hideBackground, null /* data */), requestCode);
+                hideBackground), requestCode);
     }
 
     /**
@@ -1876,7 +1855,7 @@ public final class Utils extends com.android.settingslib.Utils {
     }
 
     private static Intent getIntentForBiometricAuthentication(Resources resources,
-            int effectiveUserId, boolean hideBackground, @Nullable Intent data) {
+            int effectiveUserId, boolean hideBackground) {
         final Intent intent = new Intent();
         intent.putExtra(BIOMETRIC_PROMPT_AUTHENTICATORS,
                 BiometricManager.Authenticators.IDENTITY_CHECK);
@@ -1887,7 +1866,6 @@ public final class Utils extends com.android.settingslib.Utils {
         intent.putExtra(ChooseLockSettingsHelper.EXTRA_KEY_ALLOW_ANY_USER, true);
         intent.putExtra(EXTRA_USER_ID, effectiveUserId);
         intent.putExtra(BIOMETRIC_PROMPT_HIDE_BACKGROUND, hideBackground);
-        intent.putExtra(EXTRA_DATA, data);
         intent.setClassName(SETTINGS_PACKAGE_NAME,
                 ConfirmDeviceCredentialActivity.InternalActivity.class.getName());
         return intent;
diff --git a/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java b/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java
index 586b0fbf966..fd262f1c8e2 100644
--- a/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java
+++ b/src/com/android/settings/biometrics/fingerprint/FingerprintSettings.java
@@ -294,6 +294,8 @@ public class FingerprintSettings extends SubSettings {
                 "biometrics_authentication_requested";
         private static final String KEY_BIOMETRICS_USE_FINGERPRINT_TO_CATEGORY =
                 "biometric_settings_use_fingerprint_to";
+        private static final String KEY_GK_PW_HANDLE =
+                "gk_pw_handle";
 
         private static final int MSG_REFRESH_FINGERPRINT_TEMPLATES = 1000;
         private static final int MSG_FINGER_AUTH_SUCCESS = 1001;
@@ -348,6 +350,7 @@ public class FingerprintSettings extends SubSettings {
         private boolean mHasRunChallengeInvoker = false;
 
         private long mChallenge;
+        private long mGkPwHandle;
 
         private static final String TAG_AUTHENTICATE_SIDECAR = "authenticate_sidecar";
         private static final String TAG_REMOVAL_SIDECAR = "removal_sidecar";
@@ -637,6 +640,7 @@ public class FingerprintSettings extends SubSettings {
                         mHasFirstEnrolled);
                 mBiometricsAuthenticationRequested = savedInstanceState.getBoolean(
                         KEY_BIOMETRICS_AUTHENTICATION_REQUESTED);
+                mGkPwHandle = savedInstanceState.getLong(KEY_GK_PW_HANDLE);
             }
 
             // (mLaunchedConfirm or mIsEnrolling) means that we are waiting an activity result.
@@ -1195,6 +1199,7 @@ public class FingerprintSettings extends SubSettings {
             outState.putBoolean(KEY_HAS_RUN_CHALLENGE_INVOKER, mHasRunChallengeInvoker);
             outState.putBoolean(KEY_BIOMETRICS_AUTHENTICATION_REQUESTED,
                     mBiometricsAuthenticationRequested);
+            outState.putLong(KEY_GK_PW_HANDLE, mGkPwHandle);
         }
 
         @Override
@@ -1358,6 +1363,7 @@ public class FingerprintSettings extends SubSettings {
                 if (resultCode == RESULT_FINISHED || resultCode == RESULT_OK) {
                     runChallengeGeneratedInvokers();
                     if (BiometricUtils.containsGatekeeperPasswordHandle(data)) {
+                        mGkPwHandle = BiometricUtils.getGatekeeperPasswordHandle(data);
                         final Utils.BiometricStatus biometricAuthStatus =
                                 Utils.requestBiometricAuthenticationForMandatoryBiometrics(
                                         getActivity(),
@@ -1366,9 +1372,9 @@ public class FingerprintSettings extends SubSettings {
                         if (biometricAuthStatus != Utils.BiometricStatus.NOT_ACTIVE) {
                             Utils.launchBiometricPromptForMandatoryBiometrics(this,
                                     BIOMETRIC_AUTH_REQUEST,
-                                    mUserId, true /* hideBackground */, data);
+                                    mUserId, true /* hideBackground */);
                         } else {
-                            handleAuthenticationSuccessful(data);
+                            handleAuthenticationSuccessful(mGkPwHandle);
                         }
                     } else {
                         Log.d(TAG, "Data null or GK PW missing");
@@ -1423,7 +1429,11 @@ public class FingerprintSettings extends SubSettings {
             } else if (requestCode == BIOMETRIC_AUTH_REQUEST) {
                 mBiometricsAuthenticationRequested = false;
                 if (resultCode == RESULT_OK) {
-                    handleAuthenticationSuccessful(data);
+                    if (mGkPwHandle == 0L) {
+                        Log.e(TAG, "Gatekeeper password not set.");
+                    } else {
+                        handleAuthenticationSuccessful(mGkPwHandle);
+                    }
                 } else {
                     if (resultCode
                             == ConfirmDeviceCredentialActivity.BIOMETRIC_LOCKOUT_ERROR_RESULT) {
@@ -1479,7 +1489,7 @@ public class FingerprintSettings extends SubSettings {
             }
         }
 
-        private void handleAuthenticationSuccessful(Intent data) {
+        private void handleAuthenticationSuccessful(long gkPwHandle) {
             if (!mHasFirstEnrolled && !mIsEnrolling) {
                 final Activity activity = getActivity();
                 if (activity != null) {
@@ -1494,8 +1504,7 @@ public class FingerprintSettings extends SubSettings {
                 // Token and challenge will be updated later through the activity result
                 // of AUTO_ADD_FIRST_FINGERPRINT_REQUEST.
                 mIsEnrolling = true;
-                addFirstFingerprint(
-                        BiometricUtils.getGatekeeperPasswordHandle(data));
+                addFirstFingerprint(gkPwHandle);
             } else {
                 mFingerprintManager.generateChallenge(mUserId,
                         (sensorId, userId, challenge) -> {
@@ -1509,10 +1518,10 @@ public class FingerprintSettings extends SubSettings {
                             final GatekeeperPasswordProvider provider =
                                     new GatekeeperPasswordProvider(
                                             new LockPatternUtils(activity));
-                            mToken = provider.requestGatekeeperHat(data, challenge,
+                            mToken = provider.requestGatekeeperHat(gkPwHandle, challenge,
                                     mUserId);
                             mChallenge = challenge;
-                            provider.removeGatekeeperPasswordHandle(data, false);
+                            provider.removeGatekeeperPasswordHandle(gkPwHandle);
                             updateAddPreference();
                         });
             }
diff --git a/src/com/android/settings/password/ConfirmDeviceCredentialActivity.java b/src/com/android/settings/password/ConfirmDeviceCredentialActivity.java
index 508b833ced2..9aa5e3630de 100644
--- a/src/com/android/settings/password/ConfirmDeviceCredentialActivity.java
+++ b/src/com/android/settings/password/ConfirmDeviceCredentialActivity.java
@@ -84,7 +84,6 @@ public class ConfirmDeviceCredentialActivity extends FragmentActivity {
             "biometric_prompt_negative_button_text";
     public static final String BIOMETRIC_PROMPT_HIDE_BACKGROUND =
             "biometric_prompt_hide_background";
-    public static final String EXTRA_DATA = "extra_data";
     public static final int BIOMETRIC_LOCKOUT_ERROR_RESULT = 100;
 
     public static class InternalActivity extends ConfirmDeviceCredentialActivity {
@@ -109,7 +108,6 @@ public class ConfirmDeviceCredentialActivity extends FragmentActivity {
     private boolean mGoingToBackground;
     private boolean mWaitingForBiometricCallback;
     private int mBiometricsAuthenticators;
-    private Intent mIntentData;
 
     private Executor mExecutor = (runnable -> {
         mHandler.post(runnable);
@@ -156,7 +154,7 @@ public class ConfirmDeviceCredentialActivity extends FragmentActivity {
                         ConfirmDeviceCredentialActivity.this);
             }
 
-            setResult(Activity.RESULT_OK, mIntentData);
+            setResult(Activity.RESULT_OK);
             finish();
         }
 
@@ -205,7 +203,6 @@ public class ConfirmDeviceCredentialActivity extends FragmentActivity {
         mBiometricsAuthenticators = intent.getIntExtra(BIOMETRIC_PROMPT_AUTHENTICATORS,
                 BiometricManager.Authenticators.DEVICE_CREDENTIAL
                         | BiometricManager.Authenticators.BIOMETRIC_WEAK);
-        mIntentData = intent.getParcelableExtra(EXTRA_DATA, Intent.class);
         final String negativeButtonText = intent.getStringExtra(
                 BIOMETRIC_PROMPT_NEGATIVE_BUTTON_TEXT);
         final boolean frp =
-- 
2.55.0

